src/main/main.c

```diff
@@ -550,62 +618,79 @@ int fd_listen(re_sock_t fd, int flags, fd_h *fh, void *arg)
 			return err;
 	}
 
-#ifdef WIN32
-	/* Windows file descriptors do not follow POSIX standard ranges. */
-	i = lookup_fd_index(re, fd);
-	if (i < 0) {
-		DEBUG_WARNING("fd_listen: fd=%d - no free fd_index\n", fd);
-		return EMFILE;
+	fhs = *fhsp;
+
+	if (!fhs) {
+		fhs = mem_zalloc(sizeof(struct re_fhs), NULL);
+		if (!fhs)
+			return ENOMEM;
+
+		fhs->fd = RE_BAD_SOCK;
+		fhs->index = -1;
+
+		*fhsp = fhs;
 	}
-#else
-	i = fd;
-#endif
 
-	if (i >= re->maxfds) {
-		if (flags) {
-			DEBUG_WARNING("fd_listen: fd=%d flags=0x%02x"
-				      " - Max %d fds\n",
-				      fd, flags, re->maxfds);
-		}
-		return EMFILE;
+	DEBUG_INFO("fd_listen: fd=%d flags=0x%02x\n", fd, flags);
+
+	/* allow setting fd only once, new fd needs new fhs allocation */
+	if (fhs->fd == RE_BAD_SOCK)
+		fhs->fd = fd;
+
+	if (fhs->fd != fd) {
+		DEBUG_WARNING("fd_listen: fhs reuse conflict %d\n", fd);
+		return EBADF;
 	}
 
-	/* Update fh set */
-	if (re->fhs) {
-		re->fhs[i].fd    = fd;
-		re->fhs[i].flags = flags;
-		re->fhs[i].fh    = fh;
-		re->fhs[i].arg   = arg;
+	/* Update fhs */
+	fhs->fh	 = fh;
+	fhs->arg = arg;
+
+	/* Stop listening */
+	if (!flags && fhs->flags) {
+		fhs->fh = NULL;
+		mbuf_write_ptr(re->fhsld, (intptr_t)fhs);
+		*fhsp = mem_deref(fhs);
+		--re->nfds;
 	}
 
-	re->nfds = max(re->nfds, i+1);
+	/* Start listening */
+	if (flags && !fhs->flags) {
+		/* reference for poll loop - avoid dangling pointer */
+		mem_ref(fhs);
```

Comment:
This is why `fd_close` use the pointer location from `application` to set this storage to `NULL`, otherwise it's easy to miss this detail. after `fd_close` is called a new `re_fhs` object has to be created.

Reply:
> This is why `fd_close` use the pointer location from `application` to set this storage to `NULL`, otherwise it's easy to miss this detail.

This doesn't protect from the user copying the pointer before calling `fd_close`, so this reasoning for passing a double pointer to `fd_close` doesn't work.

But consider the use case when `fd_poll` calls two event handlers in the loop before `fhsld_flush`. One handler calls `fd_listen` without flags and another one - `fd_listen` with flags, on the same fd. Neither one calls `fd_close`, so from the user's perspective the fd is not destroyed. Yet, the first call to `fd_listen` puts it into the delete list and the other one doesn't remove it. Eventually, `fhsld_flush` destroys it and you have a dangling pointer.

Reply:
I refactored the code and `fd_listen` can not called without flags anymore.